src/test/java/com/mongodb/kafka/connect/source/schema/BsonDocumentToSchemaTest.java

```diff
@@ -0,0 +1,213 @@
+    Schema outerArrayValueSchema = outerArraySchema.valueSchema();
+    assertEquals(STRUCT, outerArrayValueSchema.type());
+
+    Schema innerArraySchema = outerArrayValueSchema.field("innerArray").schema();
+    assertEquals(ARRAY, innerArraySchema.type());
+
+    Schema innerArrayValueSchema = innerArraySchema.valueSchema();
+    assertEquals(STRUCT, innerArrayValueSchema.type());
+
+    assertEquals(INT32, innerArrayValueSchema.field("a").schema().type());
+    assertEquals(STRING, innerArrayValueSchema.field("b").schema().type());
+    assertEquals(INT32, innerArrayValueSchema.field("c").schema().type());
+  }
+
+  @Test
+  void testStructCombiningWithNulls() {
+    BsonDocument document =
+        new BsonDocument()
+            .append(
+                "outerArray",
+                new BsonArray(
+                    asList(
+                        new BsonDocument()
+                            .append("a", new BsonInt32(1))
+                            .append("b", new BsonString("foo"))
+                            .append("c", BsonNull.VALUE)
+                            .append("d", BsonNull.VALUE)
+                            .append("e", new BsonArray()),
+                        new BsonDocument()
+                            .append("a", BsonNull.VALUE)
+                            .append("b", new BsonString("foo"))
+                            .append("c", new BsonInt32(2)))));
+
+    Schema schema = inferDocumentSchema(document);
+    assertEquals(STRUCT, schema.type());
+
+    Schema outerArraySchema = schema.field("outerArray").schema();
+    assertEquals(STRUCT, outerArraySchema.valueSchema().type());
+
+    Schema outerArrayValueSchema = outerArraySchema.valueSchema();
+    assertEquals(STRUCT, outerArrayValueSchema.type());
+
+    assertEquals(INT32, outerArrayValueSchema.field("a").schema().type());
+    assertEquals(STRING, outerArrayValueSchema.field("b").schema().type());
+    assertEquals(INT32, outerArrayValueSchema.field("c").schema().type());
+    assertEquals(STRING, outerArrayValueSchema.field("d").schema().type());
+    assertNull(outerArrayValueSchema.field("d").schema().defaultValue());
+    assertEquals(STRING, outerArrayValueSchema.field("e").schema().valueSchema().type());
+    assertNull(outerArrayValueSchema.field("e").schema().valueSchema().defaultValue());
+  }
+
+  @Test
+  void testUncombinableDocumentArrayElements() {
+    BsonDocument document =
+        new BsonDocument()
+            .append(
+                "outerArray",
+                new BsonArray(
+                    asList(
+                        new BsonDocument("a", new BsonArray(asList(new BsonInt32(1)))),
+                        new BsonDocument("a", new BsonArray(asList(new BsonString("foo")))))));
+
+    Schema schema = inferDocumentSchema(document);
+    assertEquals(STRUCT, schema.type());
+
+    Schema outerArraySchema = schema.field("outerArray").schema();
+    assertEquals(STRING, outerArraySchema.valueSchema().type());
```

Comment:
This looks off as only the field "a" array type is incompatible 

So I'd expect:

```
    Schema schema = inferDocumentSchema(document);
    assertEquals(STRUCT, schema.type());

    Schema outerArraySchema = schema.field("outerArray").schema();
    assertEquals(ARRAY, outerArraySchema.type());

    Schema outerArrayValueSchema = outerArraySchema.valueSchema();
    assertEquals(STRUCT, outerArrayValueSchema.type());

    Schema fieldSchema = outerArrayValueSchema.field("a").schema();
    assertEquals(ARRAY, fieldSchema.type());
    assertEquals(STRING, fieldSchema.valueSchema().type());
```

Reply:
Is this still an issue?

Reply:
I'm thinking no, but please confirm.  I didn't really understand the problem you were pointing out.

Reply:
Take the following document:

```
{outerArray: [{a: 1}, {a: 'foo'}]}
```

The initial change set the schema to be: `{outerArray: [Type.STRING]}`
But its only the `a` field that has a type conflict: `{outerArray: [{a: Type.STRING}]}`

So the recursive change fixed that.

Reply:
OK, so that means that every struct is compatible with every other struct.